llvm/utils/TableGen/SearchableTableEmitter.cpp

```diff
@@ -426,16 +428,25 @@ void SearchableTableEmitter::emitLookupFunction(const GenericTable &Table,
     OS << "    return nullptr;\n\n";
   }
 
-  OS << "  struct KeyType {\n";
-  for (const auto &Field : Index.Fields) {
-    OS << "    " << searchableFieldType(Table, Index, Field, TypeInTempStruct)
-       << " " << Field.Name << ";\n";
+  bool ShouldReturnRange = Index.ReturnRange;
+  if (ShouldReturnRange)
```

Comment:
Why does returning a range need to change how the Key is constructed?

Reply:
std::equal_range expects both arguments to be of same type unlike in std::lower_bound where both can be of different type.

auto Idx = std::lower_bound(Table.begin(), Table.end(), Key,
    [](const **SysReg** &LHS, const **KeyType** &RHS) {
    
auto It = std::equal_range(Table.begin(), Table.end(), Key,
    [](const **SysReg** &LHS, const **SysReg** &RHS) {

Reply:
Here's a patch to fix that in a better way. https://github.com/llvm/llvm-project/pull/97784

Reply:
Thanks! Should I include those changes in my next patchset or will it be merged separately?

Reply:
You can just include them in your next patchset. Only emit the second compare function I added if `ReturnRange` is true.

Reply:
I understand that this will simplify the code changes for generating the key, however we would still need to use equal_range instead of lower_bound to return a range of answers like below. Do let me know if this is what you were expecting
```
llvm::iterator_range<const SysReg *> lookupSysRegByEncoding(uint16_t Encoding) {
  struct KeyType {
    uint16_t Encoding;
  };
  KeyType Key = {Encoding};
  struct Comp {
    bool operator()(const SysReg &LHS, const KeyType &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
    bool operator()(const KeyType &LHS, const SysReg &RHS) const {
      if (LHS.Encoding < RHS.Encoding)
        return true;
      if (LHS.Encoding > RHS.Encoding)
        return false;
      return false;
    }
  };
  auto Table = ArrayRef(SysRegsList);
  auto It = std::equal_range(Table.begin(), Table.end(), Key, Comp());
  return llvm::make_range(It.first, It.second);
}
```

Reply:
Yes that is what I was expecting.